src/main/scala/org/broadinstitute/hail/methods/IBD.scala

```diff
@@ -0,0 +1,165 @@
+package org.broadinstitute.hail.methods
+
+import org.apache.spark.SparkContext
+import org.broadinstitute.hail.variant.{Genotype, VariantDataset}
+
+object IBDInfo {
+  def apply(Z0: Double, Z1: Double, Z2: Double): IBDInfo = {
+    IBDInfo(Z0, Z1, Z2, Z1 / 2 + Z2)
+  }
+}
+
+case class IBDInfo(Z0: Double, Z1: Double, Z2: Double, PI_HAT: Double) {
+  def pointwiseMinus(that: IBDInfo): (Double, Double, Double, Double) =
+    (Z0 - that.Z0, Z1 - that.Z1, Z2 - that.Z2, PI_HAT - that.PI_HAT)
+
+  def hasNaNs: Boolean = Array(Z0, Z1, Z2, PI_HAT).exists(_.isNaN)
+}
+
+case class IBSExpectations(
+  E00: Double, E10: Double, E20: Double,
+  E11: Double, E21: Double, E22: Double = 1, nonNaNCount: Int = 1) {
+  def hasNaNs: Boolean = Array(E00, E10, E20, E11, E21).exists(_.isNaN)
+
+  def normalized: IBSExpectations =
+    IBSExpectations(E00 / nonNaNCount, E10 / nonNaNCount, E20 / nonNaNCount, E11 / nonNaNCount, E21 / nonNaNCount, E22, this.nonNaNCount)
+
+  def scaled(N: Int): IBSExpectations =
+    IBSExpectations(E00 * N, E10 * N, E20 * N, E11 * N, E21 * N, E22 * N, this.nonNaNCount)
+
+  def join(that: IBSExpectations): IBSExpectations =
+    if (this.hasNaNs)
+      that
+    else if (that.hasNaNs)
+      this
```

Comment:
These two cases are how I avoid counting IBS expectations with NaNs. In Plink, there's an if statement that achieves the same effect.